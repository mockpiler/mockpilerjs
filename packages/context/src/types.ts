export type MockContextAccessor = (key: string) => any

export interface MockContext {
  [key: string]: any
}

export type MockContextInput = MockContext | MockContextAccessor

export type ContextualizeResult = {
  template: string
  context: MockContext // proxified context
}
