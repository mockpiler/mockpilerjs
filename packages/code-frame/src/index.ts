/**
 * Expose buffer-related stuff
 */
export * from './buffer'

/**
 * Expose code-frame main class
 */
export * from './code-frame'

/**
 * Expose options
 */
export * from './options'
