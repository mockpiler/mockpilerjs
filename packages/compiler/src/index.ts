/**
 * Expose compiler main functionalities
 */
export * from './compiler'

/**
 * Expose types
 */
export * from './types'

/**
 * Expose alias
 */
export { createCompiler as mock } from './compiler'

/**
 * Expose context utils
 */
export * from '@mockpiler/context'

/**
 * Expose transformers
 */
export * from '@mockpiler/transform'
