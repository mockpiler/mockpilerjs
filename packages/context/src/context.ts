import { ContextualizeResult, MockContext, MockContextInput } from './types'
import {
  getMockContextAccessor,
  getEncodedMockContextKey,
  RawValue
} from './utils'

export function contextualizeTemplate(
  templateStrings: TemplateStringsArray,
  values: any[],
  input: MockContextInput
): ContextualizeResult {
  const accessor = getMockContextAccessor(input)

  let [template] = templateStrings
  const randomContextKey = getEncodedMockContextKey(Math.random()).slice(2)

  const rootContext: MockContext = {}
  const rootContextProxy = new Proxy(rootContext, {
    get(target, key: string) {
      return Reflect.has(target, key) ? target[key] : accessor(key)
    }
  })

  values.forEach((value, index) => {
    if (value instanceof RawValue) {
      template += `${value.raw}${templateStrings[index + 1]}`
    } else {
      const contextKey = `__mockpiler__${randomContextKey}_${getEncodedMockContextKey(
        index
      )}`

      rootContext[contextKey] = value
      template += `'${contextKey}'${templateStrings[index + 1]}`
    }
  })

  return {
    template,
    context: rootContextProxy
  }
}
