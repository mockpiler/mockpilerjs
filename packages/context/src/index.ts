/**
 * Expose context-related
 */
export * from './context'

/**
 * Expose utils
 */
export * from './utils'

/**
 * Expose context types
 */
export * from './types'
