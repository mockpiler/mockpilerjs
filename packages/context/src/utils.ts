import { MockContextAccessor, MockContextInput } from './types'

export const getEncodedMockContextKey = (n: number) => n.toString(36)

export const unknownIdent = Symbol('MockPiler.UnknownIdent')

export class RawValue {
  constructor(public raw: any) {}
}

export function raw(value: any) {
  return new RawValue(value)
}

/**
 * Utility function to get a context accessor
 * from a plain context object
 *
 * @param input
 */
export function getMockContextAccessor(
  input: MockContextInput
): MockContextAccessor {
  if (typeof input === 'function') {
    return input as any
  }

  // Generate an accessor based on given input
  return key => (key in input ? input[key] : unknownIdent)
}

/**
 * Merges given plain context objects or accessors
 *
 * @param inputs
 */
export function mergeMockContexts(
  ...inputs: MockContextInput[]
): MockContextAccessor {
  const accessors = inputs.map(getMockContextAccessor)

  return key => {
    for (const accessor of accessors) {
      const result = accessor(key)

      if (result !== unknownIdent) {
        return result
      }
    }

    return unknownIdent
  }
}
