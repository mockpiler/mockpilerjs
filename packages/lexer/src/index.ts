/**
 * Expose lexer-related things
 */
export * from './lexer'

/**
 * Expose useful constants
 */
export * from './constants'

/**
 * Expose token declarations
 */
export * from './token'
