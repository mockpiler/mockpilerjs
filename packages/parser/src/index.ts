/**
 * Expose AST nodes
 */
export * from './ast'

/**
 * Expose parser-related things
 */
export * from './parser'
