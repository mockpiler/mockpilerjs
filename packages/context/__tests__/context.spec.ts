import './test-patch'
import { contextualizeTemplate } from '../src'

const contextualizeTemplateTag = (
  template: TemplateStringsArray,
  ...values: any[]
) => contextualizeTemplate(template, values, {})

describe('context', () => {
  test('should contextualize template correctly', () => {
    const { template, context } = contextualizeTemplateTag`
  {
    key: ${'value'}
    array: [
      (2) {
        obj: ${{ n: 10 }}
      }
    ]
  }
`

    expect(template).toMatchSnapshot()

    // Note: We have to spread context since 'pretty-format'
    // has collisions with proxies
    expect({ ...context }).toMatchSnapshot()
  })
})
