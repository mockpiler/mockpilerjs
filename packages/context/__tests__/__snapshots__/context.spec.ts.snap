// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context should contextualize template correctly 1`] = `
"
  {
    key: '__mockpiler__x5lpkkh82qk_0'
    array: [
      (2) {
        obj: '__mockpiler__x5lpkkh82qk_1'
      }
    ]
  }
"
`;

exports[`context should contextualize template correctly 2`] = `
Object {
  "__mockpiler__x5lpkkh82qk_0": "value",
  "__mockpiler__x5lpkkh82qk_1": Object {
    "n": 10,
  },
}
`;
